File: src/leetcode/medium/best_time_to_buy_and_sell_stock_II/BestTimeToBuyAndSellStockII.java
package leetcode.medium.best_time_to_buy_and_sell_stock_II;

public class BestTimeToBuyAndSellStockII {
    public static void main(String[] args) {
        System.out.println(maxProfit(new int[]{7, 1, 5, 3, 6, 4}));
    }

    public static int maxProfit(int[] prices) {
        int n = prices.length;
        if (n < 2)
            return 0;
        int sellingDate = 0;
        int buyingDate = 0;
        int totalProfit = 0;
        for (int i = 1; i < n; i++) {
            if (prices[i] >= prices[i - 1])
                sellingDate++;
            else {
                totalProfit += (prices[sellingDate] - prices[buyingDate]);
                sellingDate = buyingDate = i;
            }
        }
        totalProfit += (prices[sellingDate] - prices[buyingDate]);
        return totalProfit;
    }
}

File: src/main/java/leetcode/medium/BestTimeToBuyAndSellStockII.kt
package leetcode.medium

//122. Best Time to Buy and Sell Stock II

fun main() {
    println(maxProfit(intArrayOf(7, 1, 5, 3, 6, 4)))
}

fun maxProfit(prices: IntArray): Int {
    val n = prices.size
    if (n < 2) {
        return 0
    }
    var profit = 0
    var sellDay = 0
    var buyDay = 0
    for (i in 1 until n) {
        if (prices[i] >= prices[i - 1]) {
            sellDay++
        } else {
            profit += prices[sellDay] - prices[buyDay]
            sellDay = i
            buyDay = i
        }
    }
    return profit + prices[sellDay] - prices[buyDay]
}